hoomd/Variant.h

```diff
@@ -50,27 +50,330 @@ class PYBIND11_EXPORT VariantConstant : public Variant
+        /// The value.
         Scalar m_value;
     };
 
+/** Ramp variant.
+
+    Variant that ramps linearly from A to B over a given number of steps.
+*/
+class PYBIND11_EXPORT VariantRamp : public Variant
+    {
+    public:
+
+        /** Construct a VariantRamp.
+
+            @param A The starting value.
+            @param B The ending value.
+            @param t_start The starting time step.
+            @param t_ramp The length of the ramp.
+        */
+        VariantRamp(Scalar A, Scalar B, uint64_t t_start, uint64_t t_ramp)
+            {
+            setA(A);
+            setB(B);
+            setTStart(t_start);
+            setTRamp(t_ramp);
+            }
+
+        /// Evaluate the ramp.
+        Scalar operator()(uint64_t timestep)
+            {
+            if (timestep < m_t_start)
+                {
+                return m_A;
+                }
+            else if (timestep < m_t_start + m_t_ramp)
+                {
+                // interpolate between A and B
+                uint64_t v = timestep - m_t_start;
+                double s = double(v) / double(m_t_ramp);
+                return m_B * s + m_A * (1.0 - s);
+                }
+            else
+                {
+                return m_B;
+                }
+            }
+
+        /// Set the starting value.
+        void setA(Scalar A)
+            {
+            m_A = A;
+            }
+
+        /// Get the starting value.
+        Scalar getA()
+            {
+            return m_A;
+            }
+
+        /// Set the ending value.
+        void setB(Scalar B)
+            {
+            m_B = B;
+            }
+
+        /// Get the ending value.
+        Scalar getB()
+            {
+            return m_B;
+            }
+
+        /// Set the starting time step.
+        void setTStart(uint64_t t_start)
+            {
+            m_t_start = t_start;
+            }
+
+        /// Get the starting time step.
+        uint64_t getTStart()
+            {
+            return m_t_start;
+            }
+
+        /// Set the length of the ramp.
+        void setTRamp(uint64_t t_ramp)
+            {
+            // doubles can only represent integers accuracy up to 2**53.
+            if (t_ramp >= 9007199254740992ull)
+                {
+                throw std::invalid_argument("t_ramp must be less than 2**53");
+                }
+            m_t_ramp = t_ramp;
+            }
+
+        /// Get the length of the ramp.
+        uint64_t getTRamp()
+            {
+            return m_t_ramp;
+            }
+
+    protected:
+        /// The starting value.
+        Scalar m_A;
+
+        /// The ending value.
+        Scalar m_B;
+
+        /// The starting time step.
+        uint64_t m_t_start;
+
+        /// The length of the ramp.
+        uint64_t m_t_ramp;
+    };
+
+
+/** Cycle variant
+
+    Variant that cycles linearly from A to B and back again over a given number of steps.
+*/
+class PYBIND11_EXPORT VariantCycle : public Variant
```

Comment:
Should we also have a periodic cycle where we have a phase and period and use sin or cos to cycle? Regardless, I think that this is the appropriate default cycle.

Reply:
Yes, I think a `variant.Sin` would be a good addition. I will leave that as an excessive for someone interested in using it. The hyperbolic tangent might also be nice to use as a ramp, but I agree that the simple linear interpolation methods are the most predictable defaults.